### PR TITLE
add LimitIterator class

### PR DESCRIPTION
**Pull request by javeme**
Change-Id: Iec3c3d9876b23bda0897089585b8993f7665f0ac

### CODE REVIEW

**Comment by javeme** [comment]
ci error: in `RestClientTest.testCleanExecutor()` at `Mockito.verify(pool, Mockito.never()).closeExpiredConnections();`

**Comment by javeme** [comment]
ci error: in `RestClientTest.testCleanExecutor()` at `Mockito.verify(pool, Mockito.never()).closeExpiredConnections();`

File: src/test/java/com/baidu/hugegraph/unit/iterator/FilterIteratorTest.java [hunk 1]
```diff
@@ -139,6 +138,29 @@ public void testNextWithMultiTimesWithoutAnyResult() {
...
+        Iterator<Integer> vals = list.iterator();
+
+        AtomicInteger callbackCount = new AtomicInteger(0);
+
+        Iterator<Integer> results = new FilterIterator<>(vals, val -> {
+            callbackCount.incrementAndGet();
+            return true;
+        });
+
+        Assert.assertTrue(results.hasNext());
```
Comment (Linary) [inline src/test/java/com/baidu/hugegraph/unit/iterator/FilterIteratorTest.java:156]: Move into for loop

File: src/main/java/com/baidu/hugegraph/iterator/LimitIterator.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.util.Iterator;
+import java.util.function.Function;
+
+public class LimitIterator<T> extends WrappedIterator<T> {
+
+    private final Iterator<T> originIterator;
+    private final Function<T, Boolean> filterCallback;
+
+    public LimitIterator(Iterator<T> origin, Function<T, Boolean> filter) {
```
Comment (Linary) [inline src/main/java/com/baidu/hugegraph/iterator/LimitIterator.java:30]: I think LimitIterator just pass a count is enough, this impl seems more like a ConditionIterator

**Review by Linary — CHANGES_REQUESTED** [review]

File: src/main/java/com/baidu/hugegraph/iterator/LimitIterator.java (hunk 2 above)
Comment (javeme) [reply src/main/java/com/baidu/hugegraph/iterator/LimitIterator.java:30]: actual not the same, LimitIterator stop once filterCallback returns false once

File: src/test/java/com/baidu/hugegraph/unit/iterator/FilterIteratorTest.java (hunk 1 above)
Comment (javeme) [reply src/test/java/com/baidu/hugegraph/unit/iterator/FilterIteratorTest.java:156]: for-loop is just to skip elements

File: src/main/java/com/baidu/hugegraph/iterator/LimitIterator.java (hunk 2 above)
Comment (javeme) [reply src/main/java/com/baidu/hugegraph/iterator/LimitIterator.java:30]: usage: https://github.com/hugegraph/hugegraph/blob/bc90cf9e0a636485efcd8949f665e3d8f311e1a0/hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java#L1676

**Review by Linary — APPROVED** [review]

File: src/test/java/com/baidu/hugegraph/unit/iterator/FilterIteratorTest.java [hunk 3]
```diff
@@ -139,6 +138,29 @@ public void testNextWithMultiTimesWithoutAnyResult() {
         });
     }
 
+    @Test
+    public void testNextWithOringinIteratorReturnNullElem() {
```
Comment (zhoney) [inline src/test/java/com/baidu/hugegraph/unit/iterator/FilterIteratorTest.java:142]: origin

**Review by zhoney — CHANGES_REQUESTED** [review]

**Review by zhoney — APPROVED** [review]

**Review by Linary — APPROVED** [review]

**Pull request merged**